bin/tx-prover/src/proxy/mod.rs

```diff
@@ -154,77 +179,70 @@ impl ProxyHttp for LoadBalancer {
     ///
     /// Note that the request is not removed from the queue here. It will be returned later in
     /// [Self::logging()] once the worker processes the it.
     async fn upstream_peer(
         &self,
-        session: &mut Session,
+        _session: &mut Session,
         ctx: &mut Self::CTX,
     ) -> Result<Box<HttpPeer>> {
-        if ctx.tries > 0 {
-            Self::remove_request_from_queue(Self::get_request_id(session)?).await;
-        }
+        let request_id = ctx.request_id;
 
-        // Select the worker in a round-robin fashion
-        let worker = self.lb.select(b"", 256).ok_or(Error::new_str("Worker not found"))?;
-
-        // Read request ID from headers
-        let request_id = {
-            let id = Self::get_request_id(session)?;
-            id.to_string()
-        };
-
-        // Enqueue the request ID in the worker queue
-        // We use a new scope to release the lock after the operation
-        {
-            let mut ctx_guard = QUEUES.write().await;
-            let worker_queue = ctx_guard.entry(worker.clone()).or_insert_with(Vec::new);
-
-            // Limit queue length
-            if worker_queue.len() >= self.max_queue_items {
-                Self::create_queue_full_response(session).await?;
-            }
-
-            worker_queue.push(request_id.to_string());
-        }
+        // Add the request to the queue.
+        QUEUE.enqueue(request_id).await;
 
         // Wait for the request to be at the front of the queue
         loop {
-            // We use a new scope for each iteration to release the lock
+            // We use a new scope for each iteration to release the lock before sleeping
             {
-                let ctx_guard = QUEUES.read().await;
-                if let Some(worker_queue) = ctx_guard.get(&worker) {
-                    if worker_queue[0] == request_id {
-                        break;
-                    }
-                } else {
-                    return Err(Error::new_str("Worker not found"));
+                // The request is at the front of the queue.
+                if QUEUE.peek().await.expect("Queue should not be empty") != request_id {
+                    continue;
+                }
+
+                // Check if there is an available worker
+                if let Some(worker) = self.get_available_worker().await {
+                    ctx.set_worker(worker);
+                    info!("Worker picked up the request with ID: {}", request_id);
+                    break;
                 }
+                info!("All workers are busy");
             }
             tokio::time::sleep(Duration::from_millis(100)).await;
         }
 
+        // Remove the request from the queue
+        QUEUE.dequeue().await;
+
         // Set SNI
-        let mut http_peer = HttpPeer::new(worker, false, "".to_string());
+        let mut http_peer =
+            HttpPeer::new(ctx.worker.clone().expect("Failed to get worker"), false, "".to_string());
```

Comment:
Is cloning workers relatively cheap?

Reply:
To answer this correctly, I ran a flamegraph (pasting it below). The clone does not appear to be an issue.
Let me know if you have any further concern.

![flamegraph](https://github.com/user-attachments/assets/b471d7b3-0349-4afd-8d66-98915e6d0fcb)